src/ocean/math/BinaryHistogram.d

```diff
@@ -0,0 +1,384 @@
+        17: 64ki - ∞
+
+    ***************************************************************************/
+
+    private uint[NumBins] bins;
+
+    /***************************************************************************
+
+        Struct with one uint field per bin (see this.bins), named as follows
+        (assuming Suffix == ""):
+            from_0, from_1, from_2, from_4, from_8,
+            from_16, ...,
+            from_1Ki, from_2Ki, ...,
+            etc
+
+        Useful, for example, for logging the whole histogram. An instance of
+        this struct is returned by `stats()`.
+
+    ***************************************************************************/
+
+    private struct Bins
+    {
+        import ocean.core.Traits : ctfe_i2a;
+
+        /***********************************************************************
+
+            Interprets the passed bins array as a Bins instance.
+
+            Params:
+                array = bins array to reinterpret
+
+            Returns:
+                the passed bins array as a Bins instance
+
+        ***********************************************************************/
+
+        public static Bins fromArray ( typeof(BinaryHistogram.bins) array )
+        {
+            return *(cast(Bins*)array.ptr);
+        }
+
+        /***********************************************************************
+
+            Sanity check that the offset of the fields of this struct match the
+            offsets of the elements of a BinaryHistogram.bins array. (i.e.
+            that the fromArray() function can work as intended.)
+
+        ***********************************************************************/
+
+        static assert(fieldOffsetsCorrect());
+
+        /***********************************************************************
+
+            Returns:
+                true if the offset of the fields of this struct match the
+                offsets of the elements of a BinaryHistogram.bins array
+
+        ***********************************************************************/
+
+        private static bool fieldOffsetsCorrect ( )
+        {
+            foreach ( i, field; typeof(Bins.tupleof) )
+                if ( Bins.tupleof[i].offsetof != i * BinaryHistogram.init.bins[i].sizeof )
+                    return false;
+            return true;
+        }
+
+        /***********************************************************************
+
+            CTFE generator of the fields of this struct.
+
+            Params:
+                max_power = maximum power of 2 contained within a bin
+                TODO
+
+            Returns:
+                code for the series of fields for bins up to the specified
+                maximum power of 2. (See unittest for examples.)
+
+        ***********************************************************************/
+
+        private static istring divisionBinVariables ( uint max_power,
+            istring suffix )
```

Comment:
The parameters aren't actually needed, you could use `MaxPower` and `Suffix` directly (or move the function to module scope).